File: ui/filepicker.py
"""
filepicker.py - gets and validates file path from user
"""

from pathlib import Path

from rich.console import Console
from rich.prompt import Prompt

from ui.theme import Theme


def pick_file():
    console = Console()

    # loop for file input
    while True:
        path_str = Prompt.ask(
            f"[bold {Theme.INFO}]ENTER PATH TO FILE[/bold {Theme.INFO}]"
        ).strip()
        path = Path(path_str)

        if path.is_file():
            # success message
            console.print(
                f"[{Theme.SUCCESS}]✓ File found:[/{Theme.SUCCESS}] [{Theme.DIM}]{path}[/{Theme.DIM}]\n"
            )
            return str(path)
        else:
            # error message
            console.print(
                f"[{Theme.ERROR}]✗ File not found:[/{Theme.ERROR}] {path_str}"
            )

File: src/ui/filepicker.rs
/// filepicker.rs - gets and validates file path from user

use std::path::Path;
use console::style;
use dialoguer::Input;
use crate::ui::theme::Theme;

// --- ask the user for a file path, validate it ---

pub fn pick_file() -> String {
    loop {
        let path: String = Input::new()
            .with_prompt(format!("{}", style("ENTER PATH TO FILE (or press ENTER to browse)").fg(Theme::INFO).bold()))
            .allow_empty(true)
            .interact_text()
            .unwrap_or_default();

        let path = path.trim().trim_matches('\'').trim_matches('"').to_string();

        let path = if path.is_empty() {
            if let Some(file) = rfd::FileDialog::new().pick_file() {
                file.to_string_lossy().into_owned()
            } else {
                continue;
            }
        } else {
            path
        };

        if Path::new(&path).is_file() {
            // success message
            println!(
                "{} {}",
                style("✓ File found:").fg(Theme::SUCCESS),
                style(&path).fg(Theme::DIM),
            );
            println!();
            return path;
        }

        // error message
        println!(
            "{} {}",
            style("✗ File not found:").fg(Theme::ERROR),
            path,
        );
    }
}
